### PR TITLE
[Rollup] Disallow index patterns that match rollup indices

### PR DESCRIPTION
**Pull request by polyfractal**
We should not allow the user to configure index patterns that also match rollup indices (either the index specified in the config, or previously created jobs).

Matching against rollup indices will almost certainly lead to bad behavior, because the expected fields in the user index won't exist in the rollup (due to renaming).  It is also unlikely a user will want to match against a rollup index.

For example, it is quite natural for a user to specify `metricbeat-*` as the index pattern, and then store the rollups in `metricbeat-rolled`. This will start throwing errors as soon as the rollup index is created because the indexer will try to search it.

While we're at it, we should forbid the index pattern being equal to the rollup index too.

/cc @cdahlqvist

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java [hunk 1]
```diff
@@ -336,6 +337,17 @@ public RollupJobConfig build() {
             if (indexPattern == null || indexPattern.isEmpty()) {
                 throw new IllegalArgumentException("An index pattern is mandatory.");
             }
+            if (Regex.isMatchAllPattern(indexPattern)) {
+                throw new IllegalArgumentException("Index pattern must not match all indices (as it would match it's own rollup index");
+            }
+            if (Regex.isSimpleMatchPattern(indexPattern)) {
+                if (Regex.simpleMatch(indexPattern, rollupIndex)) {
+                    throw new IllegalArgumentException("Index pattern would match rollup index name which is not allowed.");
```
Comment (jimczi) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Good catch !

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java [hunk 2]
```diff
@@ -115,6 +125,26 @@ public void onFailure(Exception e) {
         });
     }
 
+    /**
+     * Ensure that the index pattern doesn't match any pre-existing rollup indices.  We don't need to check
+     * if the pattern would match it's own rollup index because the request already validated that.
+     */
```
Comment (jimczi) [inline x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java]: Maybe we can be less strict and only forbid mixing rollup and non-rollup indices in the same pattern ?
It should be fine to create the job if the pattern matches rollup indices only. Although I wonder if the problem of mixing indices (rollup and non-rollup) is not caught by the field validation. As you mentioned they have different field names so the validation of field names should fail all the time ?

**Review by jimczi — COMMENTED** [review]
I left some comments, the index pattern should not match the rollup index name, this is an important validation. However I wonder if the second validation is really needed/possible. What happens if you define a rollup index that match an index pattern of another job ?

File: x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java (hunk 2 above)
Comment (polyfractal) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java]: Good point about the field name validation, although i'm still a bit worried about unintended consequences of matching against a rollup index.  Barring the "rollup a rollup" usecase, it doesn't seem wise to let users match against an existing rollup if we can help it?

But maybe it's not worth the extra complexity?  If it matches another rollup index, that index will exist so we can validate against the fields and it should throw an exception.  It also doesn't protect against future rollup indices from matching the defined pattern, so we'd be relying on field level validation there anyway.

I think I'm convinced.  Will make the changes, should simplify the PR a bunch :)

Comment (polyfractal) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java]: So, it turns out to be a bit more complicated.  The field-level validation is broken, due to how field caps behaves.  

Field caps only tells you the caps for the fields that match, in the resolved indices.  But it doesn't tell you about the indices that were resolved but _didn't_ match any field.  So we won't know the index pattern matched some random index without the correct fields.

That's unrelated to this change to this PR, but needs to be fixed for us to simplify.  Going to see what it takes, we may have to manually resolve the indices anyway to check.  Will keep you updated.

Comment (jimczi) [reply x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java]: Good catch. We can also add the matching indices in field caps's response. Could be per field : the list of index that doesn't have a definition for the field or global: the list of indices that match the pattern and then you can infer which index is missing from the field's response. I think it's a valid feature to add even if we don't use it in rollups.

**Comment by polyfractal** [comment]
@jimczi I pushed a commit to prune this back to just validating that the pattern doesn't self-match.  I'll follow up with the field caps limitation and see what we can do there, then circle back to Rollup and add a few tests to make sure the field-name validation works as expected after we tweak field caps.

**Review by jimczi — APPROVED** [review]
LGTM, thanks @polyfractal

**Comment by polyfractal** [comment]
Jenkins, run gradle build tests

**Pull request merged**